api/agent/agent.go

```diff
@@ -597,18 +580,33 @@ func (a *agent) runHot(ctxArg context.Context, slots chan<- slot, call *call, to
 			}
 
 			done := make(chan struct{})
-			slot := &hotSlot{done, proto, errC, container}
+			s := call.slots.queueSlot(&hotSlot{done, proto, errC, container, nil})
+
+			tryShut := false
+			tryTerm := false
+			tryIdle := false
 
 			select {
-			case slots <- slot:
+			case <-s.trigger:
 			case <-time.After(time.Duration(call.IdleTimeout) * time.Second):
-				logger.Info("Canceling inactive hot function")
-				shutdownContainer()
-				return
+				tryIdle = true
 			case <-ctx.Done(): // container shutdown
-				return
+				tryTerm = true
 			case <-a.shutdown: // server shutdown
-				shutdownContainer()
+				tryShut = true
+			}
+
+			// if we fail to eject the slot, it means that a consumer
+			// just dequeued this and acquired the slot. In that case,
+			// we will wait for that slot to complete (<-done) and
+			// perform exit/return at the beginning of this loop.
+			if (tryShut || tryTerm || tryIdle) && call.slots.ejectSlot(s) {
+				if tryIdle {
+					logger.Info("Canceling inactive hot function")
```

Comment:
same for this, we can just defer this and say hot function is shut down regardless of reason. then won't need any of these vars